(R3-091 part 2) - (CMDS) - Consolidation of Group Postal Out files into 1 file and activity entry

Consolidation of Postal Out files for the 15-20 additional hours of work. 

Currently, when you generate a postal out:

   1) user chooses the letter to generate, choose the recipients 
   2) add any additional attachments/documents 
   3) the ‘creation’ activity entry is created for that case, and x amount of documents are sent to the letter queue 
    4) when the user brings up the letter queue and selects the postal to be sent out (printed and mailed), the PDF is created for each recipient with the Cover page, original document, any attachments combined into 1 PDF/file to be printed. 

The Request is that CMDS wants the PDF that is created and then printed to be a combination of all pages to all recipients (in the selected case) into 1 document and in 1 activity entry INSTEAD OF 3 separate documents and associated activity entries.  This consolidated file and corresponding activity entry will be copied to the cases that have been selected in the new Group Number Panel implemented in R3-091

Example:   3 recipients of a Board Order document

------------------------------------------------------
Current: 
1 document containing
Cover letter/mailing insert to recipient A with name and address
Board Order document
Any extra attachments from the case/questionnaires/etc.
Within 1 activity entry, recipients name amended to the end of the activity entry

1 document containing
Cover letter/mailing insert to recipient B with name and address
Board Order document
Any extra attachments from the case/questionnaires/etc.
Within 1 activity entry, recipients name amended to the end of the activity entry

1 document containing
Cover letter/mailing insert to recipient C with name and address
Board Order document
Any extra attachments from the case/questionnaires/etc.
Within 1 activity entry, recipients name amended to the end of the activity entry


----------------------------------------------------------------
Proposed: 
1 document containing
Cover letter/mailing insert to recipient A with name and address
Board Order document
Any extra attachments from the case/questionnaires/etc.
Cover letter/mailing insert to recipient B with name and address
Board Order document
Any extra attachments from the case/questionnaires/etc.
Cover letter/mailing insert to recipient C with name and address
Board Order document
Any extra attachments from the case/questionnaires/etc.
Within 1 activity entry.  This consolidated file and corresponding activity entry will be copied to the cases that have been selected in the new Group Number Panel implemented in R3-091.

2 issues when testing:

1) Error when creating the postal out:

> Exception in thread "Thread-5" java.lang.NoClassDefFoundError: parker/serb/sql/PostalOutBulk
> 	at parker.serb.letterGeneration.LetterGenerationPanel.insertPostalOutBulk(LetterGenerationPanel.java:967)
> 	at parker.serb.letterGeneration.LetterGenerationPanel.generateLetter(LetterGenerationPanel.java:621)
> 	at parker.serb.letterGeneration.LetterGenerationPanel.lambda$processThread$3(LetterGenerationPanel.java:1209)
> 	at java.lang.Thread.run(Thread.java:745)
> Caused by: java.lang.ClassNotFoundException: parker.serb.sql.PostalOutBulk
> 	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
> 	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)


Documents Button -> Choose Document -> Generate Letter Panel -> Chose the recipients (I chose all 5 as postal, as I have been doing) -> Generate Button Clicked - > Multi-Case Docketing Panel -> Accept Button Clicked.

The document was STILL generated, but the application crashed.  The second time I tried it, it worked fine.

2) Activity Tab did NOT update until I went to another tab and came back to the Activity tab (did not change section tab or case)

--- a/src/main/java/parker/serb/bookmarkProcessing/processMailingAddressBookmarks.java
+++ b/src/main/java/parker/serb/bookmarkProcessing/processMailingAddressBookmarks.java
@@ -15,6 +15,7 @@ import parker.serb.fileOperations.WordToPDF;
 import parker.serb.sql.AdministrationInformation;
 import parker.serb.sql.CaseParty;
 import parker.serb.sql.PostalOut;
+import parker.serb.sql.PostalOutBulk;
 import parker.serb.util.FileService;
 import parker.serb.util.JacobCOMBridge;
 import parker.serb.util.StringUtilities;
@@ -25,7 +26,7 @@ import parker.serb.util.StringUtilities;
  */
 public class processMailingAddressBookmarks {
 
-    public static String processDoAEnvelopeInsert(String templatePath, String templateName, PostalOut item) {
+    public static String processDoAEnvelopeInsert(String templatePath, String templateName, PostalOut item, PostalOutBulk person) {
         String dept = StringUtilities.getDepartment();
         AdministrationInformation sysAdminInfo = AdministrationInformation.loadAdminInfo(dept);
 
@@ -85,9 +86,17 @@ public class processMailingAddressBookmarks {
         }
 
         String name = "";
-        name += item.person.trim() + System.lineSeparator();
-        if (!item.addressBlock.equals("")) {
-            name += item.addressBlock.trim();
+
+        if (person != null) {
+            name += person.person.trim() + System.lineSeparator();
+            if (!person.addressBlock.equals("")) {
+                name += person.addressBlock.trim();
+            }
+        } else {
+            name += item.person.trim() + System.lineSeparator();
+            if (!item.addressBlock.equals("")) {
+                name += item.addressBlock.trim();
+            }
         }
 
         processBookmark.process("SerbAddress", serbAddress, Document);

--- a/src/main/java/parker/serb/letterGeneration/LetterGenerationPanel.java
+++ b/src/main/java/parker/serb/letterGeneration/LetterGenerationPanel.java
@@ -51,6 +51,7 @@ import parker.serb.sql.Mediator;
 import parker.serb.sql.ORGCase;
 import parker.serb.sql.PostalOut;
 import parker.serb.sql.PostalOutAttachment;
+import parker.serb.sql.PostalOutBulk;
 import parker.serb.sql.PostalOutRelatedCase;
 import parker.serb.sql.SMDSDocuments;
 import parker.serb.util.ClearDateDialog;
@@ -74,6 +75,8 @@ public class LetterGenerationPanel extends javax.swing.JDialog {
     List<Integer> ccParties = new ArrayList<>();
     boolean sendToEmail = false;
     boolean sendToPostal = false;
+    int postalCount = 0;
+    int emailCount = 0;
     String toEmail = "";
     String ccEmail = "";
     double attachmentSize = 0.0;
@@ -485,10 +488,11 @@ public class LetterGenerationPanel extends javax.swing.JDialog {
 
     private void generateLetter() {
         List<Integer> postalIDList = new ArrayList<>();
-        getPartyList();
-
+        int emailID = 0;
         String docName;
         questionsCMDSModel answers = null;
+        
+        getPartyList();
 
         switch (Global.activeSection) {
             case "CMDS":
@@ -563,9 +567,6 @@ public class LetterGenerationPanel extends javax.swing.JDialog {
                     break;
             }
 
-            int emailID = 0;
-            int postalID = 0;
-            
             //CMDS Related Case Information (R3-091)
             List<String> relatedCasesList = new ArrayList<>();
             if (Global.activeSection.equals("CMDS")) {
@@ -599,33 +600,48 @@ public class LetterGenerationPanel extends javax.swing.JDialog {
             }
 
             if (sendToPostal) {
-                for (int i = 0; i < personTable.getRowCount(); i++) {
-                    if (!personTable.getValueAt(i, 1).equals("")
-                            && (personTable.getValueAt(i, 2).equals("Postal") || personTable.getValueAt(i, 2).equals("Both"))) {
-
-                        boolean org = false;
-                        boolean csc = false;
-                        if (personTable.getValueAt(i, 3).equals("ORG")) {
-                            org = true;
-                        } else if (personTable.getValueAt(i, 3).equals("CSC")) {
-                            csc = true;
+                //Bulk Send for Postal (R3-091)
+                if (Global.activeSection.equalsIgnoreCase("CMDS") && postalCount > 1) {
+                    int postalID = -1;
+                    for (int i = 0; i < personTable.getRowCount(); i++) {
+                        if (!personTable.getValueAt(i, 1).equals("") && (personTable.getValueAt(i, 2).equals("Postal") || personTable.getValueAt(i, 2).equals("Both"))) {
+                            
+                            //Initial Insert of the Data
+                            if (postalID <= 0){
+                                postalID = insertPostalForBulk();
+                                insertGeneratedAttachementPostal(postalID, docName);
+                                postalIDList.add(postalID);
+                                if (relatedCasesList.size() > 0) {
+                                    for (String relatedCase : relatedCasesList) {
+                                        insertPostalOutRelatedCase(postalID, relatedCase);
+                                    }
+                                }
+                            }
+                            //Adding to the Postal Out Bulk Address Table
+                            insertPostalOutBulk(postalID, personTable.getValueAt(i, 0).toString());
                         }
+                    }
+                    
+                } else {
+                    for (int i = 0; i < personTable.getRowCount(); i++) {
+                        if (!personTable.getValueAt(i, 1).equals("")
+                                && (personTable.getValueAt(i, 2).equals("Postal") || personTable.getValueAt(i, 2).equals("Both"))) {
 
-                        postalID = insertPostal(personTable.getValueAt(i, 0).toString(), org, csc);
-                        insertGeneratedAttachementPostal(postalID, docName);
-                        postalIDList.add(postalID);
-                        if (relatedCasesList.size() > 0) {
-                            for (String relatedCase : relatedCasesList) {
-                                NumberFormatService num = NumberFormatService.parseFullCaseNumberNoNGlobal(relatedCase);
+                            boolean org = false;
+                            boolean csc = false;
+                            if (personTable.getValueAt(i, 3).equals("ORG")) {
+                                org = true;
+                            } else if (personTable.getValueAt(i, 3).equals("CSC")) {
+                                csc = true;
+                            }
 
-                                PostalOutRelatedCase postalRelatedModel = new PostalOutRelatedCase();
-                                postalRelatedModel.postalOutId = postalID;
-                                postalRelatedModel.caseYear = num.caseYear;
-                                postalRelatedModel.caseType = num.caseType;
-                                postalRelatedModel.caseMonth = num.caseMonth;
-                                postalRelatedModel.caseNumber = num.caseNumber;
-
-                                PostalOutRelatedCase.insertPostalOutRelatedCase(postalRelatedModel);
+                            int postalID = insertPostal(personTable.getValueAt(i, 0).toString(), org, csc);
+                            insertGeneratedAttachementPostal(postalID, docName);
+                            postalIDList.add(postalID);
+                            if (relatedCasesList.size() > 0) {
+                                for (String relatedCase : relatedCasesList) {
+                                    insertPostalOutRelatedCase(postalID, relatedCase);
+                                }
                             }
                         }
                     }
@@ -689,11 +705,15 @@ public class LetterGenerationPanel extends javax.swing.JDialog {
             //Get Destinations
             if (personTable.getValueAt(i, 2).equals("Email")) {
                 sendToEmail = true;
+                emailCount++;
             } else if (personTable.getValueAt(i, 2).equals("Postal")) {
                 sendToPostal = true;
+                postalCount++;
             } else if (personTable.getValueAt(i, 2).equals("Both")) {
                 sendToEmail = true;
+                emailCount++;
                 sendToPostal = true;
+                postalCount++;
             }
         }
 
@@ -920,6 +940,38 @@ public class LetterGenerationPanel extends javax.swing.JDialog {
         return PostalOut.insertPostalOut(post);
     }
 
+    private int insertPostalForBulk() {
+         PostalOut post = new PostalOut();
+
+        post.section = Global.activeSection.equals("Civil Service Commission") ? Global.caseType : Global.activeSection;
+        post.caseYear = Global.caseYear;
+        post.caseType = Global.caseType;
+        post.caseMonth = Global.caseMonth;
+        post.caseNumber = Global.caseNumber;
+        post.person = "Group Send";
+        post.addressBlock = "";
+        post.userID = Global.activeUser.id;
+        post.suggestedSendDate = suggestedSendDatePicker.getText().equals("") ? null : new Date(NumberFormatService.convertMMDDYYYY(suggestedSendDatePicker.getText()));
+
+        if (Global.activeSection.equals("CMDS")) {
+            post.historyDescription = CMDSdocToGenerate.LetterName == null ? "" : CMDSdocToGenerate.LetterName;
+        } else {
+            post.historyDescription = SMDSdocToGenerate.historyDescription == null ? SMDSdocToGenerate.description : SMDSdocToGenerate.historyDescription;
+        }
+        return PostalOut.insertPostalOut(post);
+    }
+    
+    private void insertPostalOutBulk(int postalOutId, String partyID) {
+        CaseParty party = CaseParty.getCasePartyByID(partyID);
+
+        PostalOutBulk post = new PostalOutBulk();
+        post.postalOutId = postalOutId;
+        post.person = StringUtilities.buildCasePartyNameNoPreFix(party);
+        post.addressBlock = StringUtilities.buildAddressBlockforPostal(party);
+
+        PostalOutBulk.insertAddress(post);
+    }
+    
     private void insertGeneratedAttachementEmail(int emailID, String docName) {
         if (emailID > 0) {
             EmailOutAttachment attach = new EmailOutAttachment();
@@ -940,6 +992,19 @@ public class LetterGenerationPanel extends javax.swing.JDialog {
             attach.primaryAttachment = true;
             PostalOutAttachment.insertAttachment(attach);
         }
+    }
+    
+    private void insertPostalOutRelatedCase(int postalID, String relatedCase) {
+        NumberFormatService num = NumberFormatService.parseFullCaseNumberNoNGlobal(relatedCase);
+
+        PostalOutRelatedCase postalRelatedModel = new PostalOutRelatedCase();
+        postalRelatedModel.postalOutId = postalID;
+        postalRelatedModel.caseYear = num.caseYear;
+        postalRelatedModel.caseType = num.caseType;
+        postalRelatedModel.caseMonth = num.caseMonth;
+        postalRelatedModel.caseNumber = num.caseNumber;
+
+        PostalOutRelatedCase.insertPostalOutRelatedCase(postalRelatedModel);
     }
 
     private void insertExtraAttachmentsEmail(int emailID, List<Integer> postalIDList) {

--- a/src/main/java/parker/serb/letterGeneration/MultiCaseDocketingDialog.java
+++ b/src/main/java/parker/serb/letterGeneration/MultiCaseDocketingDialog.java
@@ -177,14 +177,14 @@ public class MultiCaseDocketingDialog extends javax.swing.JDialog {
             SearchTable.getColumnModel().getColumn(2).setResizable(false);
         }
 
-        DocketButton.setText("Docket");
+        DocketButton.setText("Accept");
         DocketButton.addActionListener(new java.awt.event.ActionListener() {
             public void actionPerformed(java.awt.event.ActionEvent evt) {
                 DocketButtonActionPerformed(evt);
             }
         });
 
-        CloseButton.setText("Close");
+        CloseButton.setText("None");
         CloseButton.addActionListener(new java.awt.event.ActionListener() {
             public void actionPerformed(java.awt.event.ActionEvent evt) {
                 CloseButtonActionPerformed(evt);

--- a/src/main/java/parker/serb/letterQueue/DetailedPostalOutPanel.form
+++ b/src/main/java/parker/serb/letterQueue/DetailedPostalOutPanel.form
@@ -62,11 +62,11 @@
               <EmptySpace max="-2" attributes="0"/>
               <Component id="jLabel5" min="-2" max="-2" attributes="0"/>
               <EmptySpace max="-2" attributes="0"/>
-              <Component id="jScrollPane2" min="-2" pref="125" max="-2" attributes="0"/>
-              <EmptySpace type="separate" max="-2" attributes="0"/>
+              <Component id="jScrollPane2" pref="181" max="32767" attributes="0"/>
+              <EmptySpace type="unrelated" max="-2" attributes="0"/>
               <Component id="jLabel6" min="-2" max="-2" attributes="0"/>
               <EmptySpace max="-2" attributes="0"/>
-              <Component id="jScrollPane1" pref="152" max="32767" attributes="0"/>
+              <Component id="jScrollPane1" min="-2" pref="200" max="-2" attributes="0"/>
               <EmptySpace type="separate" max="-2" attributes="0"/>
               <Group type="103" groupAlignment="3" attributes="0">
                   <Component id="cancelButton" alignment="3" min="-2" max="-2" attributes="0"/>

--- a/src/main/java/parker/serb/letterQueue/DetailedPostalOutPanel.java
+++ b/src/main/java/parker/serb/letterQueue/DetailedPostalOutPanel.java
@@ -21,6 +21,7 @@ import parker.serb.Global;
 import parker.serb.sql.CMDSCase;
 import parker.serb.sql.PostalOut;
 import parker.serb.sql.PostalOutAttachment;
+import parker.serb.sql.PostalOutBulk;
 import parker.serb.util.ClearDateDialog;
 import parker.serb.util.FileService;
 import parker.serb.util.NumberFormatService;
@@ -69,13 +70,28 @@ public class DetailedPostalOutPanel extends javax.swing.JDialog {
 
     private void loadInfo() {
         post = PostalOut.getPostalOutByID(postalID);
+        List<PostalOutBulk> postalAddressList = PostalOutBulk.getPostalOutBulkEntries(postalID);
 
         setRelatedCaseButton();
         
-        addressBlockTextArea.setText(
+        if (postalAddressList.isEmpty()){
+            addressBlockTextArea.setText(
                 (post.addressBlock.startsWith(post.person) ? "" : post.person + System.lineSeparator())
                 + post.addressBlock);
-
+        } else {
+            String AddressBlockText = "";
+            for (PostalOutBulk person : postalAddressList){
+                if (!AddressBlockText.isEmpty()){
+                    AddressBlockText += System.lineSeparator() + System.lineSeparator();
+                }
+                AddressBlockText += "          ------------ Recipient ------------" + System.lineSeparator() + System.lineSeparator();
+                AddressBlockText += (person.addressBlock.startsWith(person.person) ? "" : person.person + System.lineSeparator()) + person.addressBlock;
+            }
+            addressBlockTextArea.setText(AddressBlockText);
+        }
+        
+        addressBlockTextArea.setCaretPosition(0);
+        
         if (post.suggestedSendDate != null) {
             suggestedSendDatePicker.setText(Global.mmddyyyy.format(post.suggestedSendDate));
         }
@@ -297,11 +313,11 @@ public class DetailedPostalOutPanel extends javax.swing.JDialog {
                     .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                     .addComponent(jLabel5)
                     .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
-                    .addComponent(jScrollPane2, javax.swing.GroupLayout.PREFERRED_SIZE, 125, javax.swing.GroupLayout.PREFERRED_SIZE)
-                    .addGap(18, 18, 18)
+                    .addComponent(jScrollPane2, javax.swing.GroupLayout.DEFAULT_SIZE, 181, Short.MAX_VALUE)
+                    .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                     .addComponent(jLabel6)
                     .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
-                    .addComponent(jScrollPane1, javax.swing.GroupLayout.DEFAULT_SIZE, 152, Short.MAX_VALUE)
+                    .addComponent(jScrollPane1, javax.swing.GroupLayout.PREFERRED_SIZE, 200, javax.swing.GroupLayout.PREFERRED_SIZE)
                     .addGap(18, 18, 18)
                     .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                         .addComponent(cancelButton)

--- a/src/main/java/parker/serb/letterQueue/LetterQueuePanel.java
+++ b/src/main/java/parker/serb/letterQueue/LetterQueuePanel.java
@@ -256,6 +256,34 @@ public class LetterQueuePanel extends javax.swing.JDialog {
         }
         return true;
     }
+    
+    private void reloadActivity() {
+        if (Global.caseNumber != null) {
+            switch (Global.activeSection) {
+                case "REP":
+                    Global.root.getrEPRootPanel1().getActivityPanel1().loadAllActivity();
+                    break;
+                case "ULP":
+                    Global.root.getuLPRootPanel1().getActivityPanel1().loadAllActivity();
+                    break;
+                case "ORG":
+                    Global.root.getoRGRootPanel1().getActivityPanel1().loadAllActivity();
+                    break;
+                case "MED":
+                    Global.root.getmEDRootPanel1().getActivityPanel1().loadAllActivity();
+                    break;
+                case "Hearings":
+                    Global.root.getHearingRootPanel1().getActivityPanel1().loadAllHearingActivity();
+                    break;
+                case "Civil Service Commission":
+                    Global.root.getcSCRootPanel1().getActivityPanel1().loadAllActivity();
+                    break;
+                case "CMDS":
+                    Global.root.getcMDSRootPanel1().getActivityPanel1().loadAllActivity();
+                    break;
+            }
+        }
+    }
 
     @SuppressWarnings("unchecked")
     // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
@@ -472,6 +500,7 @@ public class LetterQueuePanel extends javax.swing.JDialog {
     private void sendButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_sendButtonActionPerformed
     if(verifySelectionSize()){
         processSendButton();
+        reloadActivity();
         loadPanelInformation();
     }
     }//GEN-LAST:event_sendButtonActionPerformed

--- a/src/main/java/parker/serb/letterQueue/LetterQueuePanel.java
+++ b/src/main/java/parker/serb/letterQueue/LetterQueuePanel.java
@@ -19,6 +19,8 @@ import parker.serb.sql.EmailOutAttachment;
 import parker.serb.sql.LetterQueue;
 import parker.serb.sql.PostalOut;
 import parker.serb.sql.PostalOutAttachment;
+import parker.serb.sql.PostalOutBulk;
+import parker.serb.sql.PostalOutRelatedCase;
 
 /**
  *
@@ -163,13 +165,7 @@ public class LetterQueuePanel extends javax.swing.JDialog {
     }
 
     private void processSendButton() {
-            //OLD Way, Single Use only
             new ConfirmationBulkDialog(Global.root, true, jTable1);
-
-//            //OLD Way, Single Use only
-//            new ConfirmationDialog(Global.root, true,
-//                    jTable1.getValueAt(jTable1.getSelectedRow(), 1).toString(),
-//                    (int) jTable1.getValueAt(jTable1.getSelectedRow(), 0));
     }
 
     private void processDeleteButton() {
@@ -178,13 +174,16 @@ public class LetterQueuePanel extends javax.swing.JDialog {
             for (int i = 0; i < jTable1.getRowCount(); i++) {
                 if (jTable1.getValueAt(i, 1).equals(true)) {
                     int rowID = (int) jTable1.getValueAt(i, 0);
+                    String sendType = jTable1.getValueAt(i, 2).toString().trim();
 
-                    if (jTable1.getValueAt(i, 2).toString().equals("Email")) {
+                    if (sendType.equals("Email")) {
                         EmailOut.removeEmail(rowID);
                         EmailOutAttachment.removeEmailAttachment(rowID);
-                    } else if (jTable1.getValueAt(i, 2).toString().equals("Postal")) {
+                    } else if (sendType.equals("Postal")) {
                         PostalOut.removeEntry(rowID);
                         PostalOutAttachment.removeEntry(rowID);
+                        PostalOutRelatedCase.deletePostalOutRelatedCaseByID(rowID);
+                        PostalOutBulk.removeEntry(rowID);
                     }
                 }
             }

--- a/src/main/java/parker/serb/letterQueue/postalSend.java
+++ b/src/main/java/parker/serb/letterQueue/postalSend.java
@@ -23,6 +23,7 @@ import parker.serb.fileOperations.WordToPDF;
 import parker.serb.sql.Activity;
 import parker.serb.sql.PostalOut;
 import parker.serb.sql.PostalOutAttachment;
+import parker.serb.sql.PostalOutBulk;
 import parker.serb.sql.PostalOutRelatedCase;
 import parker.serb.sql.RelatedCase;
 import parker.serb.util.FileService;
@@ -46,6 +47,7 @@ public class postalSend {
         // Get Letter List
         PostalOut postalEntry = PostalOut.getPostalOutByID(sendID);
         List<PostalOutAttachment> attachmentList = PostalOutAttachment.getPostalOutAttachments(sendID);
+        List<PostalOutBulk> postalAddressList = PostalOutBulk.getPostalOutBulkEntries(sendID);
 
         //Set Case Path
         String casePath = "";
@@ -69,73 +71,109 @@ public class postalSend {
                     + File.separatorChar;
         }
 
-        //Envelope Insert
-        if (!(Global.activeSection.equalsIgnoreCase("CSC") || Global.activeSection.equalsIgnoreCase("Civil Service Commission"))) {
-            //Generate Envelope Insert
-            String envelopeFileName = processMailingAddressBookmarks.processDoAEnvelopeInsert(Global.templatePath, "EnvelopeInsert.docx", postalEntry);
+        if (0 == postalAddressList.size()){
+            //Envelope Insert
+            if (!(Global.activeSection.equalsIgnoreCase("CSC") || Global.activeSection.equalsIgnoreCase("Civil Service Commission"))) {
+                //Generate Envelope Insert
+                String envelopeFileName = processMailingAddressBookmarks.processDoAEnvelopeInsert(Global.templatePath, "EnvelopeInsert.docx", postalEntry, null);
 
-            //Convert Envelope
-            String envelopeFilePDF = WordToPDF.createPDF(casePath, envelopeFileName);
+                //Convert Envelope
+                String envelopeFilePDF = WordToPDF.createPDF(casePath, envelopeFileName);
 
-            //Add Envelope To PDF Merge
-            try {
-                ut.addSource(casePath + envelopeFilePDF);
-                tempPDFList.add(casePath + envelopeFilePDF);
-                tempPDFList.add(casePath + envelopeFileName);
-            } catch (FileNotFoundException ex) {
-                SlackNotification.sendNotification(ex);
+                //Add Envelope To PDF Merge
+                try {
+                    ut.addSource(casePath + envelopeFilePDF);
+                    tempPDFList.add(casePath + envelopeFilePDF);
+                    tempPDFList.add(casePath + envelopeFileName);
+                } catch (FileNotFoundException ex) {
+                    SlackNotification.sendNotification(ex);
+                }
             }
-        }
 
-        //Convert Attachments
-        for (PostalOutAttachment attach : attachmentList) {
-            String fileName = attach.fileName;
-            String extension = FilenameUtils.getExtension(attach.fileName);
+            //Convert Attachments
+            for (PostalOutAttachment attach : attachmentList) {
+                String fileName = attach.fileName;
+                String extension = FilenameUtils.getExtension(attach.fileName);
 
-            //Convert attachments to PDF
-            //If Image
-            if (FileService.isImageFormat(fileName)) {
-                fileName = ImageToPDF.createPDFFromImage(casePath, fileName);
+                //Convert attachments to PDF
+                //If Image
+                if (FileService.isImageFormat(fileName)) {
+                    fileName = ImageToPDF.createPDFFromImage(casePath, fileName);
+                } else if (extension.equals("docx") || extension.equals("doc")) {
+                    fileName = WordToPDF.createPDF(casePath, fileName);
+                } else if ("txt".equals(extension)) {
+                    fileName = TXTtoPDF.createPDF(casePath, fileName);
+                } 
 
                 //Add Attachment To PDF Merge
                 try {
                     ut.addSource(casePath + fileName);
+                    // if NOT pdf add it to the temp list for removal
+                    if (!FilenameUtils.getExtension(fileName).equals("pdf")) {
+                        tempPDFList.add(casePath + fileName);
+                    }
+                } catch (FileNotFoundException ex) {
+                    SlackNotification.sendNotification(ex);
+                }
+            }
+
+        } else {
+            List<String> envelopePDFList = new ArrayList<>();
+            List<String> attachmentPDFList = new ArrayList<>();
+            
+            //Envelope Insert
+            if (!(Global.activeSection.equalsIgnoreCase("CSC") || Global.activeSection.equalsIgnoreCase("Civil Service Commission"))) {
+                for (PostalOutBulk postalParty : postalAddressList) {
+                    //Generate Envelope Insert
+                    String envelopeFileName = processMailingAddressBookmarks.processDoAEnvelopeInsert(Global.templatePath, "EnvelopeInsert.docx", postalEntry, postalParty);
+
+                    //Convert Envelope
+                    String envelopeFilePDF = WordToPDF.createPDF(casePath, envelopeFileName);
+
+                    //Add to Lists
+                    envelopePDFList.add(casePath + envelopeFilePDF);
+                    tempPDFList.add(casePath + envelopeFilePDF);
+                    tempPDFList.add(casePath + envelopeFileName);
+                }
+            }
+
+            //Convert Attachments
+            for (PostalOutAttachment attach : attachmentList) {
+                String fileName = attach.fileName;
+                String extension = FilenameUtils.getExtension(attach.fileName);
+
+                //Convert attachments to PDF
+                //If Image
+                if (FileService.isImageFormat(fileName)) {
+                    fileName = ImageToPDF.createPDFFromImage(casePath, fileName);
+                } else if (extension.equals("docx") || extension.equals("doc")) {
+                    fileName = WordToPDF.createPDF(casePath, fileName);
+                } else if ("txt".equals(extension)) {
+                    fileName = TXTtoPDF.createPDF(casePath, fileName);
+                } 
+
+                //Add to Lists
+                attachmentPDFList.add(casePath + fileName);
+                
+                // if NOT pdf add it to the temp list for removal
+                if (!FilenameUtils.getExtension(fileName).equals("pdf")) {
                     tempPDFList.add(casePath + fileName);
+                }
+            }
+            
+            for (String person : envelopePDFList) {
+                try {
+                    ut.addSource(person);
                 } catch (FileNotFoundException ex) {
                     SlackNotification.sendNotification(ex);
                 }
 
-                //If Word Doc
-            } else if (extension.equals("docx") || extension.equals("doc")) {
-                fileName = WordToPDF.createPDF(casePath, fileName);
-
-                //Add Attachment To PDF Merge
-                try {
-                    ut.addSource(casePath + fileName);
-                    tempPDFList.add(casePath + fileName);
-                } catch (FileNotFoundException ex) {
-                    SlackNotification.sendNotification(ex);
-                }
-                //If Text File
-            } else if ("txt".equals(extension)) {
-                fileName = TXTtoPDF.createPDF(casePath, fileName);
-
-                //Add Attachment To PDF Merge
-                try {
-                    ut.addSource(casePath + fileName);
-                    tempPDFList.add(casePath + fileName);
-                } catch (FileNotFoundException ex) {
-                    SlackNotification.sendNotification(ex);
-                }
-
-                //If PDF
-            } else if (FilenameUtils.getExtension(fileName).equals("pdf")) {
-
-                //Add Attachment To PDF Merge
-                try {
-                    ut.addSource(casePath + fileName);
-                } catch (FileNotFoundException ex) {
-                    SlackNotification.sendNotification(ex);
+                for (String attachment : attachmentPDFList) {
+                    try {
+                        ut.addSource(attachment);
+                    } catch (FileNotFoundException ex) {
+                        SlackNotification.sendNotification(ex);
+                    }
                 }
             }
         }
@@ -252,6 +290,7 @@ public class postalSend {
         PostalOut.removeEntry(sendID);
         PostalOutAttachment.removeEntry(sendID);
         PostalOutRelatedCase.deletePostalOutRelatedCaseByID(sendID);
+        PostalOutBulk.removeEntry(sendID);
 
         return casePath + savedDoc;
     }

--- a/src/main/java/parker/serb/mailLogViewer/MailLogViewerPanel.java
+++ b/src/main/java/parker/serb/mailLogViewer/MailLogViewerPanel.java
@@ -452,11 +452,11 @@ public class MailLogViewerPanel extends javax.swing.JDialog {
             if (evt.getClickCount() == 2 && !fileName.equals("")) {
                 switch (Global.activeSection) {
                     case "ORG":
-                        FileService.openFileWithORGNumber("ORG", jTable1.getValueAt(jTable1.getSelectedRow(), 2).toString(), fileName);
+                        FileService.openFileWithORGNumber("ORG", jTable1.getValueAt(jTable1.getSelectedRow(), 3).toString(), fileName);
                         break;
                     case "CSC":
                     case "Civil Service Commission":
-                        FileService.openFileWithORGNumber("CSC", jTable1.getValueAt(jTable1.getSelectedRow(), 2).toString(), fileName);
+                        FileService.openFileWithORGNumber("CSC", jTable1.getValueAt(jTable1.getSelectedRow(), 3).toString(), fileName);
                         break;
                     default:
                         FileService.openFileWithCaseNumber(Global.activeSection, caseNumber[0], caseNumber[1], caseNumber[2], caseNumber[3], fileName);

--- a/src/main/java/parker/serb/sql/PostalOutBulk.java
+++ b/src/main/java/parker/serb/sql/PostalOutBulk.java
@@ -1,0 +1,113 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package parker.serb.sql;
+
+import com.microsoft.sqlserver.jdbc.SQLServerException;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.commons.dbutils.DbUtils;
+import parker.serb.util.SlackNotification;
+
+/**
+ *
+ * @author Andrew
+ */
+public class PostalOutBulk {
+    
+    public int id;
+    public int postalOutId;
+    public String person;
+    public String addressBlock;
+    
+    public static List<PostalOutBulk> getPostalOutBulkEntries(int id) {
+        List<PostalOutBulk> emailList = new ArrayList<>();
+        
+        Statement stmt = null;
+        try {
+            stmt = Database.connectToDB().createStatement();
+
+            String sql = "SELECT * FROM PostalOutBulk WHERE postalOutID = ? ORDER BY person DESC";
+            
+            PreparedStatement preparedStatement = stmt.getConnection().prepareStatement(sql);
+            preparedStatement.setInt(1, id);
+
+            ResultSet emailListRS = preparedStatement.executeQuery();
+            
+            while(emailListRS.next()) {
+                PostalOutBulk eml = new PostalOutBulk();
+                eml.id = emailListRS.getInt("id");
+                eml.postalOutId = emailListRS.getInt("postalOutID");
+                eml.person = emailListRS.getString("person");
+                eml.addressBlock = emailListRS.getString("addressBlock");
+                emailList.add(eml);
+            }
+        } catch (SQLException ex) {
+            SlackNotification.sendNotification(ex);
+            if(ex.getCause() instanceof SQLServerException) {
+                getPostalOutBulkEntries(id);
+            } 
+        } finally {
+            DbUtils.closeQuietly(stmt);
+        }
+        return emailList;
+    }
+    
+    public static void insertAddress(PostalOutBulk item) {
+        Statement stmt = null;
+        try {
+            stmt = Database.connectToDB().createStatement();
+
+            String sql = "Insert INTO PostalOutBulk ("
+                    + "postalOutID, "
+                    + "person, "
+                    + "addressBlock "
+                    + ") VALUES (";
+                    for(int i=0; i<2; i++){
+                        sql += "?, ";   //01-02
+                    }
+                     sql += "?)";   //03
+
+            PreparedStatement preparedStatement = stmt.getConnection().prepareStatement(sql);
+            preparedStatement.setInt(1, item.postalOutId);
+            preparedStatement.setString(2, item.person);
+            preparedStatement.setString(3, item.addressBlock);
+            preparedStatement.executeUpdate();
+        } catch (SQLException ex) {
+            SlackNotification.sendNotification(ex);
+            if(ex.getCause() instanceof SQLServerException) {
+                insertAddress(item);
+            } 
+        } finally {
+            DbUtils.closeQuietly(stmt);
+        }
+    }
+    
+    public static void removeEntry(int id){
+        Statement stmt = null;
+        try {
+            stmt = Database.connectToDB().createStatement();
+
+            String sql = "DELETE FROM PostalOutBulk WHERE postalOutID = ?";
+
+            PreparedStatement preparedStatement = stmt.getConnection().prepareStatement(sql);
+            preparedStatement.setInt(1, id);
+
+            preparedStatement.executeUpdate();
+        } catch (SQLException ex) {
+            SlackNotification.sendNotification(ex);
+            if(ex.getCause() instanceof SQLServerException) {
+                removeEntry(id);
+            } 
+        } finally {
+            DbUtils.closeQuietly(stmt);
+        }
+    }
+    
+}

--- a/src/main/java/parker/serb/sql/PostalOutBulk.java
+++ b/src/main/java/parker/serb/sql/PostalOutBulk.java
@@ -33,7 +33,7 @@ public class PostalOutBulk {
         try {
             stmt = Database.connectToDB().createStatement();
 
-            String sql = "SELECT * FROM PostalOutBulk WHERE postalOutID = ? ORDER BY person DESC";
+            String sql = "SELECT * FROM PostalOutBulk WHERE postalOutID = ? ORDER BY person ASC";
             
             PreparedStatement preparedStatement = stmt.getConnection().prepareStatement(sql);
             preparedStatement.setInt(1, id);

--- a/src/main/java/parker/serb/util/SlackNotification.java
+++ b/src/main/java/parker/serb/util/SlackNotification.java
@@ -27,7 +27,6 @@ public class SlackNotification {
             message += "Method Name: " + Thread.currentThread().getStackTrace()[2].getMethodName() + "\n";
             message += "Exception Type: " + ex.getClass().getSimpleName() + "\n";
             message += "Exception Short: " + ex.toString() + "\n";
-//            message += "Stack Trace: " + convertStackTrace(ex);
 
             new Slack(Global.SLACK_HOOK)
                 .icon(Global.SLACK_ICON) // Ref - http://www.emoji-cheat-sheet.com/

